### DISCUSSION TITLE
[@types/react] The remove of onPointerEnterCapture caused unexpected errors.

### DISCUSSION THREAD

**Discussion by Thulof** [DefinitelyTyped/DefinitelyTyped#69006] [Issues with a @types package] [answered]
Yesterday, @types/react released a new version which removed onPointerEnterCapture. https://github.com/DefinitelyTyped/DefinitelyTyped/pull/68984

However, this change has resulted in unexpected errors.

# A reproduction path:
Consider a project A that uses an older version of @types/react and exports a component wrapped with forwardRef, where the component's props type is a complex type that with React.HTMLAttributes.

When this component is used in project B, which utilizes a newer version of @types/react, there will be an error indicating that the onPointerEnter property is missing.

`error TS2739: Type '{ children: string; }' is missing the following properties from type 'Pick<IDemoProps, "slot" | "style" | "title" | "type" | "key" | "defaultChecked" | "defaultValue" | "suppressContentEditableWarning" | "suppressHydrationWarning" | ... 249 more ... | "ellipsis">': onPointerEnterCapture, onPointerLeaveCapture`

# Reason:
The forwardRef function performs an `Omit<props, 'ref'>` operation on the props.

In the final type declaration file generated by project A, the Omit sometimes gets transformed into Pick. See https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/61757

If project A is using an old version of @types/react, this eventually turns into `Pick<React.HTMLAttributes, ... | 'onPointerEnterCapture' | ...>`.

Since project B is using a newer version of @types/react, where React.HTMLAttributes no longer includes onPointerEnterCapture, the onPointerEnterCapture is considered of an unknown type in the final prop types.

At this point, it becomes a required property, thus the usage of the component in project B demands that onPointerCapture must be provided.

In a project, numerous npm packages are used. If an npm package depends on an older version of @types/react, and the project that uses this package requires a newer version (for instance, the version specified for @types/react in the package.json devDependencies field is "17.x" or "latest"), then the issue described above will occur when developers simply run `pnpm i`.

[a comment by an automated account is not shown]

**Comment by Thulof** [comment]
I have created a revert PR to roll back @types/react to its previous state, in order to avoid having everyone encounter this error after updating their dependencies. https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69005

**Comment by eps1lon** [comment]
Multiple versions of the React types in a single application are not supported just like mutliple versions of React in an app are not supported.

Could you provide a full, minimal example that shows the issue?

**Reply by Thulof** [reply]
It's not an issue of multiple versions, but rather that the version of @types/react used during the build of an npm package may result in the distributed package's .d.ts files containing the string onPointerEnterCapture. 

The npm packages used by an application could have been built some time ago, which might lead to this problem. 

I apologize for not providing a minimal demo right away, i am doing, please wait a moment.

**Reply by eps1lon** [reply]
Yeah, this is another instance of TypeScript wrongfully inlining types: https://github.com/microsoft/TypeScript/issues/57779. This fix needs to happen in TypeScript. Otherwise we can't ship bug fixes.

**Reply by Thulof** [reply]
Yes, I agree with your point. However, there is already an issue with the version from yesterday, 18.2.66, and now developers around the world could easily trigger this confusing problem just by casually updating their dependencies. Can we consider reverting and publishing a 18.2.67 to minimize the impact?

Due to the TypeScript issue https://github.com/microsoft/TypeScript/issues/57779, many published npm packages' types already include code like `export declare const Foo: React.ForwardRefExoticComponent<Pick<React.HTMLAttributes, 'foo' | 'onPointerEventCapture' | '...'>>`, which is irreparable.

And now, if an application uses @types/react version 18.2.66, they will find that even if they use the Foo component correctly, TypeScript will still prompt them to pass in the onPointerEventCapture property, which can be very confusing.

**Reply by Thulof** [reply]
@eps1lon  Hi，I created a mini repo https://github.com/Thulof/types-react-pointer-issue to reproduce this issue. Please have a look， thank you.

**Reply by matej-marcisovsky** [reply]
What is the point of using DefinitelyTyped if the types it provides are wrong? Reverting this would not help anything in the long term. I believe that this is the only way that maintainers of other packages would fix this in their versions. Let them know by reporting an issue or creating a pull request. Or you can just freeze @types/react in your projects if your project is in maintenance mode, not worth the trouble, etc.

We could mark these two events as deprecated with TSDoc, but this is not the case. They are not deprecated. They never existed.

Reverting this fix would not affect me personally (I know now that this bug exists), and the final decision is not on me. I sympathize with your issues, but there are other developers that have no idea about this bug and are relying on TypeScript types, sometimes more blindly than they should.

**Reply by Thulof** [reply]
I agree with your perspective; we should adhere to correctness instead of compromising due to historical legacy issues or development costs. I am very thankful to everyone, as this has allowed me to understand the differences between maintaining an open-source project and a corporate project.

**Reply by eps1lon** [reply]
Either you manually picked `onPointerCapture` and relied on a bug, or this was TypeScript wrongfully inlining a type. We reported this before and got no buy in. There is now a dedicated issue you can subscribe to and upvote to help prioritization: https://github.com/microsoft/TypeScript/issues/57779

**Reply by eps1lon** [reply]
@Thulof It would help a lot if you can comment in https://github.com/microsoft/TypeScript/issues/57779 what code you authored and what TypeScript inlined to add another case where inlining is incorrect.

**Reply by Thulof** [reply]
@eps1lon OK, I will tidy up the previous minimal reproduction project, ensuring that only the essential parts remain, and then I will comment it in https://github.com/microsoft/TypeScript/issues/57779.

**Comment by RJM1996** [comment]
I encountered this issue a few days ago too! It turned out to be caused by an upgrade of the @types/react version, which cost me half a day!
<img width="1387" alt="C51B2B71-14E9-4F92-B244-8F08FCF02E97" src="https://github.com/DefinitelyTyped/DefinitelyTyped/assets/33800707/999bcff0-6553-437f-95bc-ad9377b2ad59">

**Comment by martymh** [comment]
👍 Seeing exactly the same type of error with a recent update to @types/react 17.0.78.

```
error TS2739: Type '{ ... }' is missing the following properties from type 'Pick<...>': onPointerEnterCapture, onPointerLeaveCapture
```

**Comment by eps1lon** [comment, marked as answer]
I'll revert this for the older type versions since those will most likely not be fixed since it would require re-publish of the affected libraries with the latest 17.x types. The change stays for React 18 types though.

Backports should be reserved for critical bug fixes and given how long we had this bug in, it doesn't seem critical.

If you're affected, please follow https://github.com/microsoft/TypeScript/issues/57779 as well.

**Comment by peatiscoding** [comment]
As for work around we are using this type augmentation.

```ts
// src/global.d.ts
import 'react'

declare module 'react' {
  interface HTMLAttributes<T> {
    onPointerEnterCapture?: (e: React.PointerEvent<T>) => void
    onPointerLeaveCapture?: (e: React.PointerEvent<T>) => void
  }
  interface RefAttributes<T> {
    onPointerEnterCapture?: (e: React.PointerEvent<T>) => void
    onPointerLeaveCapture?: (e: React.PointerEvent<T>) => void
  }
}
```

**Reply by ilia-kravtsov** [reply]
It helped me, thank you very much!

**Comment by joshualip-plaudit** [comment]
So... Is there any update on workarounds besides using type augmentation to put the removed properties back? It's extremely annoying, and [the task we're told to follow](https://github.com/microsoft/TypeScript/issues/57779) doesn't exactly help.

**Reply by Thulof** [reply]
> So... Is there any update on workarounds besides using type augmentation to put the removed properties back? It's extremely annoying, and [the task we're told to follow](https://github.com/microsoft/TypeScript/issues/57779) doesn't exactly help.

According to https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69243#pullrequestreview-1973380929, the change won't be reverted for 18 or next. 
You can override the @types/react version by using the 'resolutions' field in the package.json file.
